andreeapetcu/assignment1/Anagrams.java

```diff
@@ -0,0 +1,110 @@
+ *     ASCII characters encoding for words/sentences
+ *     For sentence anagrams punctuation marks are not considered separate words
+ */
+
+public class Anagrams {
+
+    private final int difference = 'a'- 'A';
+    private final String splitExpression = "\\s*[^a-zA-Z0-9]+\\s*";
+
+
+    private char caseInsensitive(char c){
+        if(Character.isUpperCase(c))
+            return (char) (c + difference );
+        return c;
+    }
+
+    /**
+     * Checks whether two words are anagrams of each other
+     *
+     * @param input1 first word
+     * @param input2 second word
+     * @param caseSensitive flag for case sensitive or case insensitive anagrams
+     *
+     */
+    public boolean wordAnagrams(String input1, String input2, boolean caseSensitive){
+        //contains count of each letter in the first word
+        int[] letters = new int[128];
+
+        if (input1.length() != input2.length())
+            return false;
+
+        char currentLetter;
+
+        //iterate over the first word and increase count of each letter in the word
+        for (int i=0; i< input1.length(); i++){
+            if (caseSensitive)
+                currentLetter = input1.charAt(i);
+            else
+                currentLetter = caseInsensitive(input1.charAt(i));
+
+            if (letters[currentLetter] == 0)
+                letters[currentLetter] = 1;
+            else
+                letters[currentLetter]++;
+        }
+
+        //iterate over the second word and decrease count of the letters
+        for (int i = 0; i < input2.length(); i++){
+            if (caseSensitive)
+                currentLetter = input2.charAt(i);
+            else
+                currentLetter = caseInsensitive(input2.charAt(i));
+
+            letters[currentLetter]--;
+            if (letters[currentLetter] < 0)
+                return false;
+        }
+
+        //if count of letters is not zero
+        //then the words are not anagrams
+        for (int i = 0; i < letters.length; i++){
+            if (letters[i] > 0)
+                return false;
+        }
+
+        return true;
+    }
+
+
+    /**
+     * Checks if two words/sentences are anagrams of each other
+     *
+     * @param input1 first word/sentence
+     * @param input2 second word/sentence
+     * @param caseSensitive flag for case sensitive or case insensitive anagrams
+     */
+    public boolean sentenceAnagrams(String input1, String input2, boolean caseSensitive){
+        String[] splitSentence1 = input1.split(splitExpression);
+        String[] splitSentence2 = input2.split(splitExpression);
+
+        if (splitSentence1.length != splitSentence2.length)
+            return false;
+
+        //words from sentence 1 that have a pair in sentence 2
+        boolean[] checkedWordsSentence1 = new boolean[splitSentence1.length];
+
+        //words from sentence 2 that have a pair in sentence 1
+        boolean[] checkedWordsSentence2 = new boolean[splitSentence2.length];
```

Comment:
Are you talking about cases where sentence1 = "At least he did not steal my book of fairy tales" and sentence2 = "at tales did not he least my book of steal fairy" (the word "least" has two anagrams that appear in the sentence: steal and tales)?

In this case, my algorithm looks for an anagram of "least" and the first match in the second sentence is "tales"; it now sets the corresponding position of "least" in checkWordsSentence1 to true and the corresponding position of "tales" in checkedWordsSentence2 to true. When the iteration in the first sentence gets to "steal", the first anagram in the second sentence is again "tales"; however, since "tales" is already set to true, the iteration continues and the next anagram of "steal" that is set to false in the second sentence, "least" in my example, will be the match. Should the algorithm not find a matching anagram in the second sentence for "steal", then "steal" will remain false and the for loop that checks if every entry of checkedWordsSentence1 and checkedWordsSentence2 is true, will return false.

If this is not the issue, could you please give me another example?